### PR TITLE
Script to auto-format all regions of code that a feature branch modifies

### PR DESCRIPTION
**Pull request by abelnation**
This PR introduces a new script, `./utils/codestyle/formatFeatureBranch.js` that applies an auto-formatter script to just the regions of files that you have modified between your current branch and `dev`.  The formatter script uses both `jscs` and `esformatter` to get your code as much inline with the mrdoobapproves style as possible.  

To use the script, do the following:

1) ensure you have committed all changes
2) run `./utils/codestyle/formatFeatureBranch.js`
3) (hope it doesn't give you an error ;) )
4) view formatting changes as unstaged modifications to your files
5) correct any wrong/unwanted fixes
6) commit

You should only see changes to regions of files that you have actually touched in your branch (it may bleed a line or two into neighboring regions due to how diff splits chunks).

The goal of this approach is:
1) provide folks with an automated way to apply code formatting to their changes
2) prevent PR's that modify huge parts of the code base for the sake of styling
3) help ensure that new changes abide by the style guide
4) try and enforce more rules than the mrdoobapproves webtool is able to manage.

### CODE REVIEW

**Comment by mrdoob** [comment]
PRs that modify the whole code base break other pending PRs 😕

**Comment by abelnation** [comment]
@mrdoob that's a bummer.  are you interested in having the tool integrated regardless for running on single files when need be?  do you ever do cleanup pr's like this?  or do you just view it as not really possible considering the number of pr's you have open?

Also, random aside, I'm interested in helping w/ the webgl renderer refactors that you are working on, and curious how I could help :)

**Comment by mrdoob** [comment]
I think code aesthetics tweaks is something that can be done in small PRs. Massive PRs like this one tend to bring more problems than it fixes.

> Also, random aside, I'm interested in helping w/ the webgl renderer refactors that you are working on, and curious how I could help :)

That'd be nice! I don't have any tips though, it's just a matter of following the conversation.

**Comment by abelnation** [comment]
Also, see this RE jscs being deprecated: 
https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.2rv8ji7q5

esformatter (and similar tools) are based on eslint, which seems to be the current front-runner in terms of modern javascript parsing and linting.  

If you're not interested in this direction, though, I'm happy to close this PR.

**Comment by abelnation** [comment]
Another thought: 

A test could be written that checks only files that are changed in a PR for proper formatting.  That way, you could have an automated way to give source files proper formatting, but only as each is changed and pushed to dev, preventing the change-all-things-at-once problem.

**Comment by antont** [comment]
checking PRs seems useful in any case, no matter if the whole codebase is also processed at some point or not

**Comment by bhouston** [comment]
@abelnation, maybe you can add formatter as an optional tool and undo the actual application of the reformatting to the code base?  Then we can merge it into current PRs and apply the formatting changes to them and thus ease the merging to master which also has them made.

I guess I want to see this type of stuff contributed, we just have to be creative in how to get it accepted.

**Comment by abelnation** [comment]
I like the idea proposed in a separate PR of only running the tool on the lines of files that are shown as modified by `git diff`.  I'm not sure if various tools would work well in that context, but it's certainly an interesting idea.

**Comment by abelnation** [comment]
(oops, sorry for the close/reopen.  accidental click.)
(( also, the "Close and comment" button is totally where the "Cancel" button should be))

image

File: src/animation/AnimationAction.js [hunk 1]
```diff
@@ -70,78 +69,77 @@ THREE.AnimationAction._new =
...
 
-	reset: function() {
+	reset: function () {
 
 		this.paused = false;
 		this.enabled = true;
 
-		this.time = 0;			// restart clip
-		this._loopCount = -1;	// forget previous loops
-		this._startTime = null;	// forget scheduling
```
Comment (tschw) [inline src/animation/AnimationAction.js]: All these comments were neatly aligned, and now they read worse than before. (Keep in mind the tab sized used for display on github is 8).

File: src/animation/AnimationMixer.js [hunk 2]
```diff
@@ -363,11 +357,11 @@ Object.assign( THREE.AnimationMixer.prototype, {
 			var bindings = action._propertyBindings;
 
 			// increment reference counts / sort out state
-			for ( var i = 0, n = bindings.length; i !== n; ++ i ) {
+			for ( var i = 0, n = bindings.length; i !== n; ++i ) {
 
 				var binding = bindings[ i ];
 
-				if ( binding.useCount ++ === 0 ) {
+				if ( binding.useCount++ === 0 ) {
```
Comment (tschw) [inline src/animation/AnimationMixer.js]: We usually *do* put a space between lvalue and pre/post-inc/dec operatos, I think...

**Comment by abelnation** [comment]
@tschw RE unary ops, that's one of the outstanding esformatter issues i was unable to figure out.  i filed a bug on their project about it.  

given the number of edge cases involved, i'm not sure this direction is worth pursuing, considering that the project relies pretty heavily on the mrdoobapproves tool.  does the jscs formatter script currently in the repo replicate the effects of the online tool?  if so, perhaps it's more worth it just to focus on ways to apply the jscs-based tool to lines of code that are modified by pr's

I'd prefer to close this PR if the esformatter direction is not the right one for the project.  Given that it may be challenging to replicate the mrdoobapproves rules exactly, I can't say that i believe it's worth the effort

**Comment by tschw** [comment]
It may be useful. I myself like to format the code myself. But OTOH we get submissions that are stylistically off, and it may save some work.

However, I'm not so fond about changing the entire code base. If we do it, it must be done piecemeal not to cause major breakage of pending PRs. Also, I'd like to see common sense put over strict enforcement of rules in that case. Maybe the tool should only add whitespace and never remove it?

The primary goal of a style guide is to have source code read good in a consistent fashion, where readability is far more important than strict consistency, especially in cases where the latter goal compromises the former.

Horizontal alignment is useful to make code more readable. As long as the style guide doesn't explicitly forbid it, I think it should be left alone. Line breaks are a similar issue. Looking at professional typesetting software like TeX, it takes quite an effort to automate line breaks so that they look aesthetic to the reader.

**Comment by abelnation** [comment]
esformatter allows you to specify that whitespace should remain untouched in certain cases.  

one thing I can do is create a file of sample code, with both style errors and correct style, that can be run and compared to an expected output so folks can see what to expect.  the .esformatter config file also still needs tweaking, this PR is not meant to be a final submission, but was more to have discussion around, so there is definitely room for improvement in terms of the formatting rules

**Comment by abelnation** [comment]
@bhouston @tschw ok.  i've pushed some updates to this branch for discussion.  my changes include a few things

- `codestyle.sh` runs both jscs and esformatter one after the other.  neither tool manages to catch and address all cases perfectly, but together, with the right configs, you can get closer to the desired style
- i've created code samples in `utils/codestyle/samples`.  these samples include good and bad examples, plus "expected" results and errors when running the code style script on the original file.  
- i wrote a test runner to reformat the samples and compare the results to the expected, outputting the diff for the cases that i have not yet managed to address with the tools.  

I encourage folks to check out the branch, try out the new script, and see what you think:

```
# install deps
npm install

cd utils/codestyle
# run all tests
./run_style_tests.sh

# reformat a src file
cd three.js
./utils/codestyle/codestyle src/path/to/srcfile.js
```

Outstanding issues I have not figured out yet (you can see these issues if you run the test runner script):

- arrays whose elements are "right-aligned" seem to have their trailing whitespace modified
- block comments are not made prettier

**Comment by tschw** [comment]
@abelnation 
> 1) run codestyle.sh on your modified source files

:+1: 

Possible implementation: Search the log of both the feature branch and `dev`, find the first common revision, diff the feature branch against it, and figure out the lines that were changed from its output.

> 2) run git add ... --patch on files modified in your index to "approve" each change made by the formatter (allowing you to reject the few cases where the formatter does not do the right thing)

Nice idea. When only processing modified lines, this could even be optional, figuring it would most often just do the right thing.

**Comment by abelnation** [comment]
I think that running the formatter "only on modified" lines is trickier than it sounds.  These formatters are analyzing the code in a way such that they can build a full syntax tree to analyze.  If you only submitted a portion of your sourcefile as input to the tools they would have a different tree and thus format the code differently (i.e. different amounts of indentation, etc.)  

But you could use the diff between dev and the feature to determine which chunks/files of the formatted results should be accepted or not.

**Comment by abelnation** [comment]
As far as automating this process, I think the only real option is a githook, which requires each person to set them up proactively (you cannot clone a repo and automatically have the githooks setup).  

I'm not sure it makes sense to add githooks as a process step, but perhaps just come up w/ a script to format the modified lines, and add the formatting changes as a separate commit, and have running the script be part of the guidelines for contributing changes.

**Comment by tschw** [comment]
@abelnation 
> But you could use the diff between dev and the feature to determine which chunks/files of the formatted results should be accepted or not. 

Yes!

I guess the tools will tolerate a clear cut at statement level (chopping the input at semicolons and blocks), but this is event simpler.

> As far as automating this process, I think the only real option is a githook, which requires each person to set them up proactively (you cannot clone a repo and automatically have the githooks setup). 

Doing it in the build script would work often enough, at least for in-core changes which are the most important onces ;).

**Comment by abelnation** [comment]
To me, right now, the most annoying issue I don't have a solution for yet is the modifying of array values that are right-aligned.  This is most commonly seen when aligning signed float values, e.g. this is desired:

```
var vertices = [
   1.0,  1.0, -1.0,
  -1.0, -1.0,  1.0,
   1.0,  1.0, -1.0,
  -1.0, -1.0,  1.0
];
```

but you get

```
var vertices = [
  1.0,  1.0, -1.0,
  -1.0, -1.0,  1.0,
  1.0,  1.0, -1.0,
  -1.0, -1.0,  1.0
];
```

**Comment by abelnation** [comment]
@tschw Pushed a couple new scripts:

- `filesChangedBetweenBranches.sh`: lists files that are modified between two branches, defaults to current branch and dev.  limits itself to `.js` files
- `formatBranchModifiedFiles.sh`: for each files from above script, runs formatter, then enters `git add --patch` mode on the file, for user to approve every change.  could be modified to determine file list from commit hash's or other nice presets (e.g. last commit only)

**Comment by tschw** [comment]
Scripts are going into the right direction, but I think they should become easier to use.

Actually, the correct algorithm is to look through `git log` of both `dev` and the feature branch and figure out the first commit ID both have in common. This is the state you want to diff against and should be the default behavior without any options given, so the script "just works". Anything else will get very confusing for the casual user when `dev` is pulled up, and there will be many false positives.

You can take an option to let the user specify another branch than `dev`, which is useful when you have branches that depend on one another, but I guess that goes under "nice to have".

As long as there are formatting issues like the ones you describe (and maybe there always will be), I think it's crucial that only lines that are actually edited are being formatted. Otherwise a single change to a file may easily cause quite a mess (e.g. when the user gets confused by the patched add, which is fairly likely to happen :) ).

Instead you can create a reverse patch once the script has applied its changes, remove all lines that have been changed and apply it, so everything outside the edited region is reverted.

Pseudo-code:
```
1. check that all changes are properly checked in, abort if not
2. get logs of the feature branch and the base branch (`dev` by default)
3. figure out first common commit id in both logs, let's call it base-ref
4. diff against base-ref and figure out all lines that have been changed
5. run the formatter
6. reverse diff against head
7. remove all lines that were changed (according to the first diff)
8. patch to revert changes outside the edited region
9. add all files that were changed & commit
```

**Comment by abelnation** [comment]
@tschw 

I think that your git log approach is covered by simply using the name of your branch as the pointer to the current commit and the latest commit on dev (provided you have pulled the latest from upstream/dev).  

The script currently defaults to `dev...$(git rev-parse --abbrev-ref HEAD)` (i.e., the diff between latest commit on dev, and latest commit on your current branch).  It simply allows you to specify other branches or git commit hashes if you like.  I can easily restrict those options.  

> Recipe:
> 1. check that all changes are properly checked in, abort if not

got it

> 2. get logs of the feature branch and the base branch (dev by default)
> 3. figure out first common commit id in both logs, let's call it base-ref
> 4. diff against base-ref and figure out all files / lines that have been changed

The current approach is for (2-4) is below.  It lists the file names of any file shown as modified between latest dev and the feature branch.  It does not show line numbers, as that requires displaying a full diff.  Adding line numbers to the script would require some external work ([e.g.](http://stackoverflow.com/questions/9848347/can-i-make-git-diff-only-show-the-changed-file-names-and-line-numbers))

```
git --no-pager diff --name-only dev..$(git rev-parse --abbrev-ref HEAD) \
    | grep -E ".*\.js$"
```

> 5. run the formatter on the files that have been changed

got it

> 6. reverse diff against head
> 7. remove all lines that were changed (according to the first diff)
> 8. patch to revert changes outside the edited region
> 9. add all files that were changed & commit

This is the step I'm the least clear on how to execute.  i'm not sure what the best/simplest method would be to filter the format diff down to only formatted chunks whose line numbers were included in the file diff referenced in the diff in (4)

**Comment by mrdoob** [comment]
The problem I see with these automatic formatters is with the fact that, even if a formatting guide of sorts exist, sometimes there are exceptions for code prettiness.

The typical case is something like this:

```js
array[ i     ] = 1;
array[ i + 1 ] = 2;
array[ i + 2 ] = 3;
```

As far as I'm aware, tools like esformater are not able to do that. Currently, when the code gets reformated ends up like this:

```js
array[ i ] = 1;
array[ i + 1 ] = 2;
array[ i + 2 ] = 3;
```

Some time the workaround is doing this:

```js
array[ i + 0 ] = 1;
array[ i + 1 ] = 2;
array[ i + 2 ] = 3;
```

Which is not ideal...

This is just one code example, there are others that I can't think of right now.

There are also files that come from other repos that shouldn't be reformatted (ie. glft).  So yeah, even if it's tempting to write a tool that automates everything I tend to prefer doing it manually, case by case.

**Comment by abelnation** [comment]
@mrdoob agreed.  I think it's futile to assume a tool/script can get it right 100% (even mrdoobapproves does not correct all issues), hence the ideas/scripts around how to confirm/accept/reject each of the formatting changes (see above).  i'm hesitant to put too much more energy into this PR, as I think the discussion should be around ideal process (if one exists), not the execution. 

Some key questions in my mind are:

- Do you want to try and provide a tool that contributors can invoke to help them format to your style guide? or do you simply want them to do their best?
- Can you rely on contributors to make the right calls, style wise?
- OR are you interested in a tool designed to be used by "primary" maintainers to bring contributions into consistency w/ the style guide, and to help keep the code-base consistent as it evolves.
- OR none of the above

I find these tools/formatters fun to work with, probably because I have some mild form of OCD, but I don't want to sink too much time into a solution that won't really provide that much real value.

**Comment by tschw** [comment]
@abelnation 
> The script currently defaults to dev...$(git rev-parse --abbrev-ref HEAD) (i.e., the diff between latest commit on dev, and latest commit on your current branch).

Ah, good then! I didn't get that from your description and did not read into the code.

> Adding line numbers to the script would require some external work

Yep, I was assuming that the diff would be parsed somehow.

> This is the step I'm the least clear on how to execute (specifically step 8). i'm not sure what the best/simplest method would be to filter the format diff down to only formatted chunks whose line numbers were included in the file diff referenced in the diff in (4)

Same here. There are many ways to tweak the diffs produced by git, so you may find some that make it easier. And yes, there is a remaining little bit of challenge involved, but no, I don't think it's anywhere near rocket science. I guess you may find a parser for node.js - it's a standardized format.

As for step 8, `patch` is a POSIX tool that applies a diff produced by `diff` (or `git diff`, for that matter). I'm pretty sure you can apply a diff (= perform a patch) just using git, although I don't know how to invoke it for that purpose off the top of my head.

I think that when you can ensure that only edited lines are formatted, this tool could be of great value. You can simply code in your style of choice, then run the tool to bring your feature basically in shape. By common sense, you or a maintainer may want to add an extra space here and there, but starting with differently formatted code that would still save most of the work that will otherwise have to be done manually.

"Hey, your feature doesn't look formatted according to our style guide, please run `npm prettify-branch`" (or whatever the final invocation would be called)...

**Comment by abelnation** [comment]
@tschw I think I have an approach that sounds viable, similar to your steps you outlined but clarifies the last few "hand-wavey" steps ;) maybe you were envisioning these steps yourself and I just didn't parse your recipe fully.

1. check that all changes are properly checked in, abort if not
2. pull latest from dev
3. `git diff dev..<current-branch>`, parsing modified line ranges from output (I've got a simple script that effectively does this now)
5. run the formatter on the files that have been changed
6. `git diff` to see changes made by formatter.  
7. use output from (6) to generate a patch file, and filter out modified chunks from diff that are not in the line ranges found in (3)
8. `git reset` or `git checkout .` to revert changes made by formatter
9. apply patch generated in (7) to files
10. commit changes

**Comment by mrdoob** [comment]
> Do you want to try and provide a tool that contributors can invoke to help them format to your style guide? or do you simply want them to do their best?

At the moment they have a (online) tool already. As far as I understand this is a command line approach of that tool.

> Can you rely on contributors to make the right calls, style wise?

I'm surprised of how well contributors tend to adjust to the formatting. At the same time, I like to be open in case someone does a different style in some specific case that I like.

I'm open for variations and experimentation. I think it would be a shame if people were forced to use a tool that would censor their style suggestions.

> OR are you interested in a tool designed to be used by "primary" maintainers to bring contributions into consistency w/ the style guide, and to help keep the code-base consistent as it evolves.

Having linter-jscs installed in Atom seems to be pretty much all I need myself.

But if you guys think this is handy, I'm open. I just wouldn't like to have mass PRs with mass re-stylings.

**Comment by mrdoob** [comment]
Not super happy with having `.esformatter` in the root though...

**Comment by abelnation** [comment]
@mrdoob @tschw, well this has been a silly exercise ;)

I have a complete, functional script that does the above described workflow.  User flow is:

1) ensure you have committed all changes
2) run `./utils/codestyle/formatFeatureBranch`
3) (hope it doesn't give you an error ;) )
4) view formatting changes as unstaged modifications to your files

You should only see changes to regions of files that you have actually touched in your branch (it may bleed a little into neighboring regions due to how diff splits chunks).  You can then inspect the formatting diff, correct any weird formattings, and commit as a formatting update.

@mrdoob `.esformatter` has been moved into `../utils/codestyle/`

**Comment by tschw** [comment]
@abelnation 
> maybe you were envisioning these steps yourself and I just didn't parse your recipe fully.

Well, I was describing the reverse logic: Letting the formatter just go over the whole file and revert what's not in the edited region by applying a filtered `git diff -R`.

> 1. check that all changes are properly checked in, abort if not
> 2. pull latest from dev
> 3. git diff dev..<current-branch>, parsing modified line ranges from output (see: utils/codestyle/linesChangedBetweenBranches.sh)

Might be outdated by now, but I would not pull up dev automatically. Also you'd have to rebase the feature branch to dev for this approach to work reliably (which would be even worse). The "reverse parse" you used in a previous post seemed to be the more appropriate way of figuring out what the feature branch changes.

> You should only see changes to regions of files that you have actually touched in your branch 

Great news!

> it may bleed a little into neighboring regions due to how diff splits chunks.

Why is it so hard to just parse it exactly?

AFAICT you'd just have to turn all `+` entries into a list of line numbers and use that list to filter the other diff reading through it with a similar parser. Even without support from a library it should be fairly easy to do within node. You can narrow the parse to just one case using precise format options creating the diffs.

**Comment by abelnation** [comment]
@tschw for some reason, when I limited the diff to zero extra lines of context around the changed lines, the resulting filtered diff was failing when I tried to apply it.  Adding 1 line of extra context solved the problem.  

> Well, I was describing the reverse logic: Letting the formatter just go over the whole file and revert what's not in the edited region by applying a filtered git diff -R.

This is essentially what happens.  I figure out which files are changed between dev and the feature.  I then run the formatter over the entire files for that list.  I generate a diff of the formatting step w/ the previous commit.  I then filter that diff to just the regions that were touched by the feature branch.  I then revert the full formatting changes and then apply the filtered diff.  Perhaps I miscommunicated?  

I'm sure that this tool could be tweaked and fine-tuned ad nauseum, however, I can't honestly say I'm very interested in taking it much further.  This idea has turned into quite a project that, when compared to other features, does not seem worth this much time.  If this sounds like a useful tool to add to the repo, great.  If others are interested in adding to this, fantastic.  If it will require hours more work, I can't say I'm interested, and would rather just close this.

**Comment by tschw** [comment]
It's gotten quite a bit of well-written code and it'd be really sad if it was thrown away! However, changing lines that were not edited makes it risky to use in practice (= recommend to use it on a stylistically off PR within a review). Even if you abandon working on it, I'm still in support for merging the code.

@abelnation 
> For some reason, when I limited the diff to zero extra lines of context around the changed lines, the resulting filtered diff was failing when I tried to apply it. Adding 1 line of extra context solved the problem.

Why does filtering have to happen at chunk level? And why is context being processed?

AFAICT,  an edit within a diff is undone by simply changing the deleting lines to context and removing the re-adding ones.

If because the formatter messes with line breaks, it might be better to just prevent that from happening. They are trouble for many reasons with automatic formatting anyway. If that behavior is kept, I believe it's still doable - although admittedly trickier to get the filtering right :).

**Comment by abelnation** [comment]
@tschw I suppose I'm trying to balance effectiveness for simplicity of execution.  The output of `git diff -r` consists of chunks that differ, e.g. a range of line numbers followed by the lines changed.  I could get more involved and literally rewrite the diff (e.g. to modify the line number ranges as I modify which line modifications are included), but it just feels like an uneven complexity / value ratio.  

I also worry about how to filter diff output within the individual chunks.  Currently, I filter the diff output on a per-chunk basis, which is fairly straightforward and without edge cases.  Once you start mucking w/ the actual content of each chunk, you get into all sorts of complexity wrt which lines to include/exclude, how to modify the line ranges etc.

**Comment by tschw** [comment]
@abelnation 
> I could get more involved and literally rewrite the diff

Ahm... Yes, that's what I've been suggesting all along. Maybe the term "filtering" obscured it. What I meant is to parse the diff into a data structure, filter that, then format it back into a textual representation. Just like stuff like this is normally done. 

> fairly straightforward and without edge cases

Which edge cases are you referring to?

Track input line numbers while parsing, track output line numbers while emitting. That should be it. Still fairly straightforward and shouldn't be too much of a journey from where your code's at.

In case the diff codec comes out nicely, it would be a useful small-sized node module in its own right - even if limited to "unified" format.

> but it just feels like an uneven complexity / value ratio

I think you are underestimating the value of running a transform on changed code only. IMHO it's of great value, greater than just automatic formatting and its applicability is not at all limited to this project:

Good names in code often take several attempts to become good. Making a commit from a regular expression applied to all changes in a feature branch would be super practical in this case. Other examples are linting, smaller formatting transforms such as removal of trailing whitespace or retabbing.

**Pull request closed without merging**